RFD 19 - Event Fetch API with Pagination (#5885)
gravitational/teleport: rfd/0019-event-iteration-api.md

## Details 3

The current API for event fetch and search is over (https://github.com/gravitational/teleport/blob/master/lib/auth/apiserver.go#L246) plain HTTPS and not GRPC and runs on the authentication server.

I propose implementing a new GRPC API that replaces and deprecates the old HTTP API endpoints. The old API endpoints should be kept as is but documentation should mention that it is deprecated and advise clients to use the new API.

### GRPC Messages 3.1

Out-of-session event request.
This replaces and deprecates the `GET /:version/events` HTTP endpoint.

These should be defined in [api/client/proto/authservice.proto](https://github.com/gravitational/teleport/blob/master/api/client/proto/authservice.proto).

```protobuf
message GetEventsRequest {
   // Namespace, if not set, defaults to 'default'
   Namespace string
   // Oldest date of returned events
   StartDate Timestamp
   // Newest date of returned events
   EndDate Timestamp
   // EventType is optional, if not set, returns all events

>    EventType string

Reviewer: right now the query is only limited to event types which is not useful by itself. Can we also search on event `JSON text` so people can find anything within that field? I think you can do that in dynamo DB.

Some use-case:

1. Want to find everything that happened on X server.
2. Want to find everything that had Y label.
3. Want to find all user Z events. 

I am suggesting to add a `Query` parameter that will be evaluated against the JSON text. This is in addition to `EventType`

Reviewer 2: I'd push this out of scope. In DynamoDB this would activate full table scan and it will be timing out. Instead  I think in V2 we should really build a proper backend, like splunk or ES and revisit the query parameter.

Author: Agreed. This is way too much work to get working on current and future backends without some serious changes to schema too I believe. There's a lot of issues to figure out and architect around that I feel are vastly out of scope for what this RFD intends to fix and to deliver in a reasonable timeframe.

   // Maximum amount of events returned
   Limit int64
   // When supplied the search will resume from the last key
   StartKey string
}
```

In-session event request.
This replaces and deprecates the `GET /:version/events` HTTP endpoint.

```protobuf
message GetSessionEventsRequest {
   // SessionID is a required valid session ID
   SessionID string
   EventType string
   Limit int64
   StartKey string
}
```

Event response from both endpoints.

```protobuf
message Events { 
   Items repeated oneof Event
   // the key of the last event if the returned set did not contain all events found i.e limit < actual amount. this is the key clients can supply in another API request to continue fetching events from the previous last position

>    LastKey string

Reviewer: I wonder if we can include a `total` field to the response?  The support for it will vary between data sources but we can return `-1` when not supported.

Author: Technically we can do it but for pretty much all backends it would slow down queries a fair bit. In my opinion we can have the client perform this calculation if needed.

Reviewer: What do you mean by client performing these calculations?

Reviewer: Technically we can implement everything:) but realistically we should be clear that total amount will not be supported then. Which is fine if everyone agrees.

Author: I'm assuming `total` here is the amount of events returned by this message and all events in the date selection preceding. Figuring out that number means that we have to at least select and count all events preceding the `StartKey` associated with this request which significantly removes performance benefits that we get by pagination since we would rescan events over and over again.

Author: I.E unless I am missing something, calculating `total` would place `O(n)` stress on the datastore.

Author: If client repeatedly fetches events, it should be able to calculate total by summing the all amounts of events it has received previously assuming `StartKey` is always the previous `LastKey`.

Reviewer: There 2 things:
1. Scanning the events, running actual query in the DB.
2. Returning the results which may be gigabytes of data. 

Both of them addressed differently. In here we are mostly focusing on #2 as this is the major bottleneck at the moment. Running an actual query on pre-processed by backend data might be faster then we think but I guess it's out of scope.

Author: Running the actual query in the DB isn't as bad as returning data but it can cause an explosion in load on very busy days since every request will have to rescan events on the backend as is currently without some form of additional state tracking. This is a solvable problem but I think that addressing it is out of scope.

Reviewer 3: Based on my read it does sound like a to solve later problem if we look at ES or Splunk etc type backends. The way splunk works for example is data is pulled to a search head to keep state and the client keeps a cursor into the data. There search also works on a streaming concept, so in the UI you can start to interact with results while it's still building the results if doing a really large search.

It's a really great system, but seems way more complicated than this design is attempting to target.

#### Backends 3.3.2

Performed some investigative work on how we can implement this pagination support in the different backends. The standard solution that works with the current backend interface is to simply query all events in a given date range from the backend and then filter out the paginated section using the given start key and limit. This is however something we want to avoid as it involves a lot of uneeded processing on every request by the authentication server.

Pagination support will needed to be added to the `IAuditLog` interface defined in `lib/events/api.go`. This involves adding a start key function parameter and an optional last key parameter in the return value of the `SearchEvents` and `SearchSessionEvents` methods.

> Pagination in DynamoDB seems possible to implement using start keys (partition + sort key) and global indexes for requesting only the paginated events and should be compatible with our current format.

Reviewer 4: Can you provide an example of the DynamoDB query you will be using?

   // the key of the last event if the returned set did not contain all events found i.e limit < actual amount. this is the key clients can supply in another API request to continue fetching events from the previous last position
   LastKey string
}
```

#### Notes on the EventType parameter 3.1.1

Currently the DynamoDB implementation is incorrect in the HTTP API as described in [#5435](https://github.com/gravitational/teleport/issues/5435). This should be fixed so that event filtering works correctly in the new GRPC API although I'd like some discussion on if it should be fixed in the current API or should be left as is.

### GRPC Endpoints 3.2

This RFD calls for two new endpoints. One endpoint for out-of-session event requests and one for in-session event requests.

These endpoints will live on the `AuthService` GRPC service in [api/client/proto/authservice.proto](https://github.com/gravitational/teleport/blob/master/api/client/proto/authservice.proto) and are defined as follows:

```proto
rpc GetEvents(GetEventsRequest) returns (Events);
```

```proto
rpc GetSessionEvents(GetSessionEventsRequest) returns (Events);
```

### Implementation details 3.3

#### Auth service 3.3.1

> Handling of these new GRPC endpoints will be implemented in [lib/auth/grpcserver.go](https://github.com/gravitational/teleport/blob/master/lib/auth/grpcserver.go). `auth.Client.SearchEvents` and `auth.Client.SearchSessionEvents` will be refactored with pagination support implemented with the same startkey/lastkey interface as seen in the GRPC API. These methods are what the GRPC handlers call. The current HTTP API implementation will be slightly altered to deal with these changes but will remain identicatal in functionality, simlpy not exposing the new internal pagination support.

Reviewer 5: ```suggestion
Handling of these new GRPC endpoints will be implemented in [lib/auth/grpcserver.go](https://github.com/gravitational/teleport/blob/master/lib/auth/grpcserver.go). `auth.Client.SearchEvents` and `auth.Client.SearchSessionEvents` will be refactored with pagination support implemented with the same startkey/lastkey interface as seen in the GRPC API. These methods are what the GRPC handlers call. The current HTTP API implementation will be slightly altered to deal with these changes but will remain identical in functionality, simply not exposing the new internal pagination support.
```

Pagination support will needed to be added to the `IAuditLog` interface defined in `lib/events/api.go`. This involves adding a start key function parameter and an optional last key parameter in the return value of the `SearchEvents` and `SearchSessionEvents` methods.

Pagination in DynamoDB seems possible to implement using start keys (partition + sort key) and global indexes for requesting only the paginated events and should be compatible with our current format.

Not an expert on the format used for filelog but since it is a plain sequence of JSON documents, implementing efficient pagination seems difficult. On this backend we might need to fall back to loading all events into memory and then sorting out the events that fall within the range of events requested.

> Firestore seems to have good support for pagination with it's query cursors and document snapshots which allow us to define a range between startkey and startkey + limit. This means we can efficiently query a subset of events.

Reviewer 5: ```suggestion
Firestore seems to have good support for pagination with its query cursors and document snapshots which allow us to define a range between startkey and startkey + limit. This means we can efficiently query a subset of events.
```

#### Client library

> The Go client library has to be updated to work with this new API. It currently uses the HTTP API
> and should be switched over completely to the new GRPC API as we deprecate the HTTP API. Adding pagination support to the existing methods in the library is not possible without a breaking change and unless that is acceptable, then adding variants of the event fetch methods with the `Paginated` suffix seems like the best option.

Reviewer 5: ```suggestion
The Go client library has to be updated to work with this new API. It currently uses the HTTP API and should be switched over completely to the new GRPC API as we deprecate the HTTP API. Adding pagination support to the existing methods in the library is not possible without a breaking change and unless that is acceptable, then adding variants of the event fetch methods with the `Paginated` suffix seems like the best option.
```

Pagination support will needed to be added to the `IAuditLog` interface defined in `lib/events/api.go`. This involves adding a start key function parameter and an optional last key parameter in the return value of the `SearchEvents` and `SearchSessionEvents` methods.

Pagination in DynamoDB seems possible to implement using start keys (partition + sort key) and global indexes for requesting only the paginated events and should be compatible with our current format.

Not an expert on the format used for filelog but since it is a plain sequence of JSON documents, implementing efficient pagination seems difficult. On this backend we might need to fall back to loading all events into memory and then sorting out the events that fall within the range of events requested.

Firestore seems to have good support for pagination with it's query cursors and document snapshots which allow us to define a range between startkey and startkey + limit. This means we can efficiently query a subset of events.

> As for the etcd backend, the official Go client now has support for API v3 which has good pagination support with the `RangeRequest` message and initial investigation shows that implementing pagination in our backend using may require some refactoring but is realistic.

Reviewer 2: We don't store events in etcd, so this is not necessary

Author: Thank, will remove and add a section I missed to commit on sqlite support.

Author: 👍 We don't appear to store events in sqlite either do we? Only dynamodb and firestore right?

Reviewer 5: I think the other supported backends for events are `stdout` and `file://` (possibly also `syslog`)

Author: You can't query events from `stdout` and `syslog` can you? How would that work?

Reviewer 5: Sorry, perhaps 'backend' was the wrong term here. We support writing events to all the listed destinations under `audit_events_uri` in the [config reference](https://goteleport.com/teleport/docs/config-reference/), which are `dynamodb`, `firestore`, `file` and `stdout`. Indeed, Teleport uses `file` out of the box as the default so presumably has some kind of rudimentary pagination implementation over that which should be included.

I've no idea what happens if you set the only destination for `audit_events_uri` as `stdout` - I presume Teleport refuses to start, as the web UI wouldn't be able to query it.

> Issue [#5885](https://github.com/gravitational/teleport/pull/5885) about implementing a MySQL event backend is also relevant to this discussion. If we were to directly implemented a MySQL compatible backend, we could perform pagination by giving each event a 64 bit integer number that increases sequentially. Event search would then be performed by filtering first by date and then selecting a range of event ID's. With a B-Tree index on dates and event ID's this would allow efficient pagination. If we go with an API/plugin based approach as @knisbet suggested, we could effectively allow for the same thing by exposing a search endpoint that in turn does this date and event ID filtering.

Reviewer 2: I would push those out of scope of this PR, because it's not clear if we actually are going to support MySQL as a part of this RFD implementation (I assume not)

Author: Yeah for sure, just wanted to mention it and how we could do it but I could also remove it if we should address this separately later.

Author: Removed
